### ISSUE TITLE
"npm test" doesn't work on Windows

### ISSUE THREAD

**Issue by ysmood** [petkaantonov/bluebird#465]
Branch  7fde23fabd6f6f4a
Windows 7, system language is Japanese. Error info:
```bash
$ npm test

> bluebird@2.9.1 test c:\Users\ys\Desktop\bluebird
> node tools/test.js

Error: assert.js:1:14 Non-ASCII character:
 (U+000D)
    at c:\Users\ys\Desktop\bluebird\tools\utils.js:19:23
    at Array.forEach (native)
    at Object.checkAscii (c:\Users\ys\Desktop\bluebird\tools\utils.js:12:30)
    at eval (eval at getTaskFunction (c:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413:33), <anonymous>:5:27)
From previous event:
    at eval (eval at getTaskFunction (c:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413:33), <anonymous>:4:31)
    at promise.then.finally.then.process.send.type (c:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:507:32)
    at process._tickCallback (node.js:419:13)
From previous event:
    at Object.actions.newTask (c:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:500:22)
    at Object._onImmediate (c:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:546:39)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
__stackCleaned__ true
name Error
message assert.js:1:14 Non-ASCII character:
 (U+000D)
npm ERR! Test failed.  See above for more details.
npm ERR! not ok code 0
```

By the way, why not add a CI for Windows? Such as https://ci.appveyor.com/.

**Comment by petkaantonov** [comment]
This isn't npm test not working on windows (I am windows user myself). The problem is you have somehow added the carriage return (`U+000D`) character to the `async.js` file and this fails the code quality check (no source files may contain such characters).

You need to configure your text editor to use unix (LF) line endings instead of windows. This is also specified in the `.editorconfig` file of the project root.

**Comment by ysmood** [comment]
@petkaantonov Sorry, I changed the git setting for CRLF. Now `npm install` goes wrong:

```bash
C:\Users\ys\Desktop\bluebird>npm i
\
> bluebird@2.9.1 prepublish C:\Users\ys\Desktop\bluebird
> node tools/build.js --no-debug --main --zalgo --browser --minify

npm ERR! Error: write EIO
npm ERR!     at errnoException (net.js:904:11)
npm ERR!     at Object.afterWrite (net.js:720:19)
npm ERR! If you need help, you may report this *entire* log,
npm ERR! including the npm and node versions, at:
npm ERR!     <http://github.com/npm/npm/issues>

npm ERR! System Windows_NT 6.1.7601
npm ERR! command "C:\\Program Files\\nodejs\\\\node.exe" "C:\\Program Files\\nod
ejs\\node_modules\\npm\\bin\\npm-cli.js" "i"
npm ERR! cwd C:\Users\ys\Desktop\bluebird
npm ERR! node -v v0.10.30
npm ERR! npm -v 1.4.21
npm ERR! syscall write
npm ERR! code EIO
npm ERR! errno EIO
npm ERR!
npm ERR! Additional logging details can be found in:
npm ERR!     C:\Users\ys\Desktop\bluebird\npm-debug.log
npm ERR! not ok code 0
```

**Comment by petkaantonov** [comment]
From quick search it seems such errors could be  caused by [wrong code pages](http://stackoverflow.com/a/21184384/995876). Anyway this isn't really a place to discuss such problems as all watchers of bluebird will get email notifications from the issue tracker. We have a [mailing list](https://groups.google.com/forum/#!forum/bluebird-js) or you can come to the IRC channel #promises @freenode

**Comment by ysmood** [comment]
@petkaantonov Yes. I figured it out. Now still the `npm test` doesn't pass.

![qq20150126-2](https://cloud.githubusercontent.com/assets/1415488/5893809/6c559612-a52d-11e4-8481-8fe96e648bae.jpg)

```bash
Possibly unhandled AssertionError: false == true
    at assertErrorHasLongTraces (C:\Users\ys\Desktop\bluebird\test\mocha\api_exceptions.js:8:5)
    at C:\Users\ys\Desktop\bluebird\test\mocha\api_exceptions.js:91:17
    at process._tickCallback (node.js:419:13)
^--- With additional stack trace: Error: safe_error
    at Suite.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:29:21)
    at context.describe.context.context (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\interfaces\bdd.js:74:10)
    at Object.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:27:1)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:185:27
    at Array.forEach (native)
    at Mocha.loadFiles (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:182:14)
    at Mocha.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:394:31)
    at eval (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413:33), <anonymous>:8
3:19)
    at Promise.each.then.error (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413
:33), <anonymous>:82:16)
From previous event:
    at Context.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:56:57)
    at Test.Runnable.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runnable.js:218:15)
    at Runner.runTest (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:374:10)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:452:12
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:299:14)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:309:7
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:248:23)
    at Object._onImmediate (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:276:5)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
^--- With additional stack trace: Error: safe_error
    at Suite.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:29:21)
    at context.describe.context.context (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\interfaces\bdd.js:74:10)
    at Object.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:27:1)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:185:27
    at Array.forEach (native)
    at Mocha.loadFiles (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:182:14)
    at Mocha.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:394:31)
    at eval (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413:33), <anonymous>:8
3:19)
    at Promise.each.then.error (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413
:33), <anonymous>:82:16)
From previous event:
    at Context.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:56:57)
    at Test.Runnable.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runnable.js:218:15)
    at Runner.runTest (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:374:10)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:452:12
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:299:14)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:309:7
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:248:23)
    at Object._onImmediate (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:276:5)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
^--- With additional stack trace: Error: safe_error
    at Suite.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:29:21)
    at context.describe.context.context (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\interfaces\bdd.js:74:10)
    at Object.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:27:1)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:185:27
    at Array.forEach (native)
    at Mocha.loadFiles (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:182:14)
    at Mocha.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:394:31)
    at eval (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413:33), <anonymous>:8
3:19)
    at Promise.each.then.error (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413
:33), <anonymous>:82:16)
From previous event:
    at Context.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:56:57)
    at Test.Runnable.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runnable.js:218:15)
    at Runner.runTest (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:374:10)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:452:12
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:299:14)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:309:7
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:248:23)
    at Object._onImmediate (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:276:5)
    at processImmediate [as _immediateCallback] (timers.js:345:15)

Some tests failed:
    api_exceptions.js For additional details you can run it individually  with the command `node tools/test --run=api_ex
ceptions.js`

noStackPrint true
__stackCleaned__ true
name Error
message Some tests failed:
    api_exceptions.js For additional details you can run it individually  with the command `node tools/test --run=api_ex
ceptions.js`

stdout
stderr Possibly unhandled AssertionError: false == true
    at assertErrorHasLongTraces (C:\Users\ys\Desktop\bluebird\test\mocha\api_exceptions.js:8:5)
    at C:\Users\ys\Desktop\bluebird\test\mocha\api_exceptions.js:91:17
    at process._tickCallback (node.js:419:13)
^--- With additional stack trace: Error: safe_error
    at Suite.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:29:21)
    at context.describe.context.context (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\interfaces\bdd.js:74:10)
    at Object.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:27:1)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:185:27
    at Array.forEach (native)
    at Mocha.loadFiles (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:182:14)
    at Mocha.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:394:31)
    at eval (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413:33), <anonymous>:8
3:19)
    at Promise.each.then.error (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413
:33), <anonymous>:82:16)
From previous event:
    at Context.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:56:57)
    at Test.Runnable.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runnable.js:218:15)
    at Runner.runTest (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:374:10)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:452:12
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:299:14)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:309:7
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:248:23)
    at Object._onImmediate (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:276:5)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
^--- With additional stack trace: Error: safe_error
    at Suite.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:29:21)
    at context.describe.context.context (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\interfaces\bdd.js:74:10)
    at Object.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:27:1)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:185:27
    at Array.forEach (native)
    at Mocha.loadFiles (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:182:14)
    at Mocha.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:394:31)
    at eval (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413:33), <anonymous>:8
3:19)
    at Promise.each.then.error (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413
:33), <anonymous>:82:16)
From previous event:
    at Context.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:56:57)
    at Test.Runnable.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runnable.js:218:15)
    at Runner.runTest (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:374:10)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:452:12
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:299:14)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:309:7
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:248:23)
    at Object._onImmediate (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:276:5)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
^--- With additional stack trace: Error: safe_error
    at Suite.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:29:21)
    at context.describe.context.context (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\interfaces\bdd.js:74:10)
    at Object.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:27:1)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:185:27
    at Array.forEach (native)
    at Mocha.loadFiles (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:182:14)
    at Mocha.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\mocha.js:394:31)
    at eval (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413:33), <anonymous>:8
3:19)
    at Promise.each.then.error (eval at getTaskFunction (C:\Users\ys\Desktop\bluebird\tools\job-runner\job-runner.js:413
:33), <anonymous>:82:16)
From previous event:
    at Context.<anonymous> (C:\Users\ys\Desktop\bluebird\test\mocha\done.js:56:57)
    at Test.Runnable.run (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runnable.js:218:15)
    at Runner.runTest (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:374:10)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:452:12
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:299:14)
    at C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:309:7
    at next (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:248:23)
    at Object._onImmediate (C:\Users\ys\Desktop\bluebird\node_modules\mocha\lib\runner.js:276:5)
    at processImmediate [as _immediateCallback] (timers.js:345:15)

npm ERR! Test failed.  See above for more details.
npm ERR! not ok code 0
```

**Comment by petkaantonov** [comment]
That's an issue with the test runner in node 10.30 (node 10.34 works fine), if you run it individually `node tools/test --run=api_exceptions` the test passes.

The fix is to specify api_exceptions test file as requiring its own isolated process, however I removed that as it slows down the runner and latest node doesn't have issues.

**Comment by ysmood** [comment]
THX. That works! But I have a project that depends on bluebird. When use bluebird `v2.8.0`, its unit tests will pass on both Linux and Windows, but for `> v2.80`, only Windows will fail, the error info is related to `promise join`. Project link https://github.com/ysmood/nofs. You can run the test yourself. The CI page is https://ci.appveyor.com/project/ysmood/nofs.

Sample error info:
```bash
     Uncaught TypeError: Cannot call method 'join' of undefined
    at reconstructStack (C:\Users\ys\Desktop\nofs\node_modules\bluebird\js\main\captured_trace.js:96:27)
    at CapturedTrace.attachExtraTrace (C:\Users\ys\Desktop\nofs\node_modules\bluebird\js\main\captured_trace.js:88:19)
    at Promise._attachExtraTrace (C:\Users\ys\Desktop\nofs\node_modules\bluebird\js\main\debuggability.js:92:19)
    at Promise._rejectCallback (C:\Users\ys\Desktop\nofs\node_modules\bluebird\js\main\promise.js:437:10)
    at C:\Users\ys\Desktop\nofs\node_modules\bluebird\js\main\promise.js:452:17
    at C:\Users\ys\Desktop\nofs\src\utils.coffee:15:7
    at Object.oncomplete (evalmachine.<anonymous>:107:15)
```

**Comment by petkaantonov** [comment]
Can you pull `reconstruct-stack-bug` branch and try with the commit from there applied? Or if you have a minimal test case that reproduces that is good too.

**Comment by ysmood** [comment]
Fixed on `v2.9.3`, and I don't know why. But after all thx!
